qdb/tracer.py

```diff
@@ -450,6 +502,90 @@ def set_quit(self):
+
+        outexc = None
+        outmsg = None
+        with capture_output() as (out, err), \
+                self._new_execution_timeout(code), \
+                self.inject_default_namespace() as stackframe:
+            try:
+                if not repr_fn and not pprint:
+                    self.eval_fn(
+                        code,
+                        stackframe,
+                        'single',
+                    )
+                else:
+                    try:
+                        # Do some some custom single mode magic that lets us
+                        # call the repr function on the last expr.
+                        value = progn(
+                            code,
+                            self.eval_fn,
+                            stackframe,
+                        )
+                    except QdbPrognEndsInStatement:
+                        # Statements have no value to print.
+                        pass
+                    else:
+                        if pprint:
+                            value = pformat(value)
+                        if repr_fn:
+                            value = repr_fn(value)
+                        print(value)
+            except Exception as e:
+                outexc = type(e).__name__
+                outmsg = self.exception_serializer(e)
+            else:
+                outmsg = out.getvalue().rstrip('\n')
+
+        if outexc is not None or outmsg is not None:
+            self.cmd_manager.send_print(code, outexc, outmsg)
+
+        self.update_watchlist()
+
+    def _stack_jump_to(self, index):
+        """
+        Jumps the stack to a specific index.
+        Raises an IndexError if the desired index does not exist.
+        """
+        # Try to jump here first. This could raise an IndexError which will
+        # prevent the tracer's state from being corrupted.
+        self.curframe = self.stack[index][0]
+
+        self.curindex = index
+        self.curframe_locals = self.curframe.f_locals
+        self.update_watchlist()
+
+    def stack_shift_direction(self, direction):
+        """
+        Shifts the stack up or down depending on dirction.
+        If direction is positive, travel up, if direction is negative, travel
+        down. If direction is 0, do nothing.
+        If you cannot shift in the desired direction, an IndexError will be
+        raised.
+        """
+        if direction == 0:
+            return  # nop
+
+        stride = -1 if direction > 0 else 1
+        stack = self.stack
+        stacksize = len(stack)
+        curindex = self.curindex
+        skip_fn = self.skip_fn
+        target = None
+        while 0 < curindex < stacksize:
+            curindex += stride
+            if not skip_fn(stack[curindex][0].f_code.co_filename):
+                target = curindex
+                break
+
+        if target is None:
+            raise IndexError('Shifted off the stack')
+        self._stack_jump_to(curindex)
```

Comment:
It's guaranteed that target == curindex here, but I think it's clearer to use target.